### ISSUE TITLE
Design pattern: singleton prototype and builder

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3437]
Design pattern: singleton, prototype and builder<br>
<br>
http://bit.ly/2tui5tg<br>
<br>
Christophe<br>
<br>
<div>
<div>
<p>&nbsp;</p>
<p>In my previous article, I spoke about the factory patterns. These patterns are part of creational patterns. In this post we&rsquo;ll focus on the rest of the creational patterns: Singleton, Builder and Prototype.</p>
<p>In my opinion, these patterns are less important than factories. Yet, it&rsquo;s still useful to know them. Following the same logic as my previous article, I&rsquo;ll provide an UML description, simple java examples (so that even if you don&rsquo;t know java you can understand) and present real examples from famous Java frameworks or APIs.</p>
<p>I&rsquo;ll sometimes use factories so <a href="http://coding-geek.com/design-pattern-factory-patterns/">read my previous article</a> if you don&rsquo;t feel comfortable with factory patterns.</p>
<p>&nbsp;</p>
<h1><span id="Creational_Patterns">Creational Patterns</span></h1>
<p>Creational patterns are design patterns that deal with object initialization and overcome the limitations of constructors. The <strong>Gang of Four</strong> in their book <strong>&ldquo;</strong><em>Design Patterns: Elements of Reusable Object-Oriented Software&rdquo;</em> described five of them:</p>
<ul>
<li>Singleton,</li>
<li>Builder,</li>
<li>Prototype,</li>
<li>Abstract Factory,</li>
<li>Factory pattern.</li>
</ul>
<p>Since this&nbsp;book was&nbsp;released (in 1994), many creational patterns have been&nbsp;invented:</p>
<ul>
<li>other type of factories (like the static one),</li>
<li>pool pattern,</li>
<li>lazy initialization,</li>
<li>dependency injection,</li>
<li>service locator,</li>
<li>&hellip;</li>
</ul>
<p>In this post, we&rsquo;ll only focus on the rest of the GoF&rsquo;s creational patterns I haven&rsquo;t already described. As I said in the introduction, they are less important than factories because you can live without them (whereas factories are the backbone of many applications and frameworks). But, they are useful and unlike factories they don&rsquo;t make the code much more difficult to read.</p>
<p>&nbsp;</p>
<h1><span id="Singleton_Pattern">Singleton Pattern</span></h1>
<p>This pattern is the most famous. During the last decades, it was over-used but its popularity has decreased since. I personally avoid &nbsp;using it since it makes the code <strong>more difficult to unit test</strong> and creates a <strong>tight coupling</strong>. I prefer to use a factory (like the Spring container) that deals with the number of authorized instances of a class, we&rsquo;ll speak about this approach. I think you should avoid the singleton pattern. In fact, the most important use of this pattern is to be able to answer an interviewer when he asks &ldquo;what is a singleton?&rdquo;. This&nbsp;pattern is very controversial and there are still people in favor of it.</p>
<p>That being said, according to the GoF a singleton aims to:</p>
<blockquote>
<p><em>&ldquo;Ensure a class only has one instance, and provide a global point of access to it&rdquo;</em></p>
</blockquote>
<p>So, there are 2 requirements for a class to be a singleton:</p>
<ul>
<li>Having a unique instance</li>
<li>Being accessible from anywhere</li>
</ul>
<p>Some people only think about the first requirement (like me few years ago). In this case, the class is only a <strong>single instance</strong>.</p>
<p>Let&rsquo;s look how to do a singleton in UML</p>
<p><a href="http://coding-geek.com/wp-content/uploads/2015/06/singleton.png"><img alt="singleton pattern" src="http://coding-geek.com/pagespeed_static/1.JiBnMqyl6S.gif"></a></p>
<p>&nbsp;</p>
<p>In this UML diagram, the Singleton class has 3 items:</p>
<ul>
<li>a class attribute (instance): this attribute contains the unique instance of the singleton class.</li>
<li>a class public method (getInstance()) : it provides the only way to get the unique instance of the class Singleton. The method can be called from anywhere since it&rsquo;s a class method (and not an instance method).</li>
<li>a private constructor (Singleton()) : it prevents anyone to instantiate a Singleton with a constructor.</li>
</ul>
<p>In this example, a developer that needs an instance of Singleton will call the Singleton.getInstance() class method.</p>
<p>The singleton instance inside the Singleton class can be:</p>
<ul>
<li>pre-initialized&nbsp;(which means it is instantiated before someone call getInstance())</li>
<li>lazy-initialized (which means it is instantiated during the first call of getInstance())</li>
</ul>
<p>Of course a real singleton has other methods and attributes to do its business logic.</p>
<p>&nbsp;</p>
<h2><span id="Java_implementations">Java implementations</span></h2>
<p>Here is a very simple way to create a singleton in Java using the pre-instantiated approach.</p>
<pre>
public class SimpleSingleton {
 private static final SimpleSingleton INSTANCE = new SimpleSingleton();
 
 private SimpleSingleton() { }
 
 public static SimpleSingleton getInstance(){
 return INSTANCE;
 }
 }
 
</pre>
<p>Using this way, the singleton instance is created only once when the class is loaded by the classloader. If the class is never used in your code, the instance won&rsquo;t be instantiated (because the classloader of the JVM&nbsp;won&rsquo;t load it) and therefore waste memory. But if the class appears and&nbsp;you don&rsquo;t use it (for example if&nbsp;it&rsquo;s only used in a very very rare condition), the singleton will be initialized for nothing. Unless your singleton takes a huge amount of memory, you should use this way.</p>
<p>Still, if you need to create your singleton only when it&rsquo;s really used (the lazy initialization), here is a way to do it in a multithreaded environment. This part is a bit tricky since it involves thread coherency.</p>
<pre>

 public class TouchySingleton {
 private static volatile TouchySingleton instance;
 
 private TouchySingleton() {
 }
 
 public static TouchySingleton getInstance() {
 if (instance == null) {
 synchronized (TouchySingleton.class) {
 if (instance == null) {
 instance = new TouchySingleton();
 }
 }
 }
 return instance;
 }
 }
 
</pre>
<p>As I said, it&rsquo;s really more difficult to read (this is why the pre-instanciated way is better). This singleton involves a lock to avoid that 2 threads calling the getInstance() at the same time create 2 instances. Since the lock is costly, there is first a test without a lock then a test with the lock (it&rsquo;s a double-checked locking) so that when the instance already exists the lock is not used.</p>
<p>Other particularity, the instance has to be volatile to ensure that its state is&nbsp;the same on the different processor cores when it is created.</p>
<p>&nbsp;</p>
<h2><span id="When_do_you_need_to_use_a_singleton">When do you need to use a singleton?</span></h2>
<ul>
<li>When you need only one resource (a database connection, a socket connection &hellip;)</li>
<li>To avoid multiple instances of a stateless class to avoid memory waste</li>
<li>For business reasons</li>
</ul>
<p><strong>You shouldn&rsquo;t use singleton for sharing variables/data between different objects since it produces a very tight coupling!</strong></p>
<p>&nbsp;</p>
<h2><span id="Why_shouldnt_you_use_a_singleton">Why shouldn&rsquo;t you use a singleton?</span></h2>
<p>At the beginning, I said that you shouldn&rsquo;t use singletons because of the way you get the singleton. It&rsquo;s based on a class function that can be called anywhere in the code. &nbsp;I read an excellent answer <a href="http://stackoverflow.com/questions/137975/what-is-so-bad-about-singletons">on stackoverflow</a> that gives 4 reasons why it&rsquo;s bad:</p>
<ul>
<li>With singletons, you hide the dependencies between the classes instead of exposing them through the interfaces. This means you need to read the code of each method to know if a class is using another class.</li>
<li>They violate the&nbsp;<a href="https://en.wikipedia.org/wiki/Single_responsibility_principle">single responsibility principle</a>: they control their own creation and lifecycle (using lazy initialization, the Singleton chose when it is created). A class should only focus on what it is meant to do. If you have a Singleton that manages people, it should only manage people and not how/when it is created.</li>
<li>They inherently cause code to be tightly&nbsp;coupled. This makes faking or mocking them for unit testing very difficult.</li>
<li>They carry states around for the lifetime of the application (for stateful singletons).
<ul>
<li>It makes unit testing difficult since you can end up with a situation where tests need to be ordered which is a piece of nonsense. By definition, each unit test should be independent from each&nbsp;other.</li>
<li>Moreover, it makes the code less predictable.</li>
</ul>
</li>
</ul>
<p>Ok, so singleton are&nbsp;bad. But what should you use instead?</p>
<p>&nbsp;</p>
<h2><span id="Use_a_single_instance_instead_of_a_singleton">Use a single instance instead of a singleton</span></h2>
<p>A singleton is just a specific type of single instance that can be getting anywhere with its class method. If you remove this second requirement, you remove many problems. But how can you deal with single instances?</p>
<p>A possible way is to manage single instances with a factory and Dependency Injection (it will be the subject of a future post).</p>
<p>Let&rsquo;s take an example to understand:</p>
<ul>
<li>You have a PersonBusiness class that needs a unique DatabaseConnection instance.</li>
<li>Instead of using a singleton to get this connection, the PersonBusiness will have a DatabaseConnection attribute.</li>
<li>This attribute will be injected at the instantiation of PersonBusiness by its constructor. Of course, you can inject any type of DatabaseConnection:
<ul>
<li>A MysqlDatabaseConnection for your development environment</li>
<li>A OracleDatabaseConnection for the production environment</li>
<li>A MockDatabaseConnection for the unit tests</li>
</ul>
</li>
<li>At this stage, nothing prevents the DatabaseConnection to be unique. This is where the factory is useful. You delegate the creation of PersonBusiness to a factory and this factory also takes care of the creation of DatabaseConnection:
<ul>
<li>It chooses which kind of connection to create (for example using a property file that specify the type of connection)</li>
<li>It ensures that the DatabaseConnection is unique.</li>
</ul>
</li>
</ul>
<p>&nbsp;</p>
<p>If you didn&rsquo;t understand what I&rsquo;ve just said, look the next java example then re-read this part again, it should by more comprehensive. Otherwise, feel free to tell me.</p>
<p>Here is an example in Java where the factory creates a MysqlDatabaseConnection but you could imagine a more complex factory that decides&nbsp;the type of connection according to a property file or an environment variable.</p>
<pre>
////////////////An interface that represents a database connection
 public interface DatabaseConnection {
 public void executeQuerry(String sql);
 }
 
 ////////////////A concrete implementation of this interface
 ////////////////In this example it's for a mysql database connection
 public class MysqlDatabaseConnection implements DatabaseConnection {
 public MysqlDatabaseConnection() {
 // some stuff to create the database connection
 }
 
 public void executeQuerry(String sql) {
 // some stuff to execute a SQL query
 // on the database
 }
 }
 
 ////////////////Our business class that needs a connection
 public class PersonBusiness {
 DatabaseConnection connection;
 
 //dependency injection using the constructor
 // it is a singleton because the factory that
 //creates a PersonBusiness object ensure that
 //UniqueDatabaseConnection has only one instance
 PersonBusiness(DatabaseConnection connection){
 this.connection = connection;
 }
 
 //a method that uses the injected singleton
 public void deletePerson(int id){
 
 connection.executeQuerry("delete person where id="+id);
 }
 }
 
 ////////////////A factory that creates business classes
 //////////////// with a unique MysqlDatabaseConnection
 public class Factory {
 private static MysqlDatabaseConnection databaseConnection = new MysqlDatabaseConnection();
 
 public static MysqlDatabaseConnection getUniqueMysqlDatabaseConnection(){
 return databaseConnection;
 }
 
 public static PersonBusiness createPersonBusiness(){
 //we inject a MysqlDataConnection but we could chose
 //another connection that implements the DatabaseConnection
 //this is why this is a loose coupling
 return new PersonBusiness(databaseConnection);
 }
 
 }
 
</pre>
<p>This is not a good example since the PersonBusiness could have a single instance since it has no state. But you could imagine that there are a ContractBusiness and a HouseBusiness that also needs that unique DatabaseConnection.</p>
<p>Still, I hope you see that using dependency injection + a factory you end up with a single instance of DatabaseConnection in your business classes as if you used a singleton. But this time, it&rsquo;s a loose coupling, which means instead of using a MysqlDatabaseConnection, you can easily use&nbsp;a MockDatabaseConnection for&nbsp;testing only the PersonBusiness class.</p>
<p>Moreover, it&rsquo;s easy&nbsp;to know that&nbsp;PersonBusiness is using a&nbsp;DatabaseConnection. You just have to look at the attributes of the class and not one of the 2000 lines of code of the class (ok, imagine this class has many functions and the overall takes 2000 lines of code).</p>
<p>This approach is used by most Java framework (Spring, Hibernate&hellip;) and Java containers (EJB containers). It&rsquo;s not a real singleton since you can instantiate the class multiple times if you want to and you can&rsquo;t get the instance from everywhere. But if you only create your instances through the factory/container, you&rsquo;ll end up with a unique instance of the class&nbsp;in your code.</p>
<p>Note: I think the Spring Framework is very confusing because its &ldquo;singleton&rdquo; scope is only a single instance. It took me some time to understand that it wasn&rsquo;t a real GoF&rsquo;s singleton.</p>
<p>&nbsp;</p>
<h2><span id="Some_thoughts">Some thoughts</span></h2>
<p>The single instance has the same drawback than the singleton when it comes to global states. <strong>You should avoid using a single instance to share data between different classes!</strong> The only exception I see is caching:</p>
<ul>
<li>Imagine you have a trading application that makes hundreds of calls per seconds and it only needs to have the stock prices from the last minutes. You could use a single instance&nbsp;(StockPriceManager) shared among the trading business classes, and every function that needs the prices would get it from the Cache. If the price is outdated, the cache would refresh it. In this situation, the drawbacks of tight coupling are worth the gain in performance. But when you need to understand a bug in production because of this global state you cry (I&rsquo;ve been there and it wasn&rsquo;t funny).</li>
</ul>
<p>&nbsp;</p>
<p>I told you to use the single instance&nbsp;approach instead of the singleton but sometimes it is worth using the real singleton when you need this object in all of the classes. For example when you need to log:</p>
<ul>
<li>Each class needs to log and this log class is often unique (because the logs are written in the same file). Since all classes use the log class, you know that every class has an implicit dependency to this log class. Moreover, it&rsquo;s not a business needs so it&rsquo;s &ldquo;less important&rdquo; to unit test the logs (shame on me).</li>
</ul>
<p>Writing a singleton is easier than writing a single instance using Dependency Injection. For a quick and dirty solution I&rsquo;d use a singleton. For a long and durable solution I&rsquo;d use a&nbsp;single instance. Since most applications are based on frameworks, the implementation of the single instance is easier than from scratch (assuming you know how to use the framework).</p>
<p>If you want to know more about singletons:</p>
<p>&nbsp;</p>
<h2><span id="Real_examples">Real examples</span></h2>
<p>The single instance&nbsp;pattern uses a factory.&nbsp;If you use an instantiable factory, you might need to ensure that this factory is unique. And more broadly, when you use a factory you might want it to be unique in order to avoid that 2 factory instances mess with each other. You could use a &ldquo;meta-factory&ldquo; to build the unique factory but you&rsquo;ll end up with the same problem for the &ldquo;meta-factory&rdquo;. So, the only way to do that is to create the factory with a singleton.</p>
<p>It&rsquo;s the case for <a href="http://docs.oracle.com/javase/8/docs/api/java/awt/Toolkit.html">java.awt.Toolkit</a>&nbsp; in the old graphical library AWT. This class provides a getDefaultToolkit() method that gives the unique&nbsp;Toolkit instance and&nbsp;it&rsquo;s the only way to get one. Using this toolkit (which is a factory), you can create a windows, a button, a checkbox &hellip;</p>
<p>&nbsp;</p>
<p>But you can also encounters singletons for other concerns. When you need to monitor a system in Java, you have to use the class java.lang.Runtime. I guess this class has to be unique because it represents the global state (environment variables)&nbsp;of the process. If I quote the <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Runtime.html">java API</a>:</p>
<blockquote>
<p>&ldquo;Every Java application has a single instance of class Runtime that allows the application to interface with the environment in which the application is running. The current runtime can be obtained from the <strong>getRuntime</strong> method.&rdquo;</p>
</blockquote>
<p>&nbsp;</p>
<h1><span id="Prototype_Pattern">Prototype Pattern</span></h1>
<p>I&rsquo;ve used prototypes through Spring but I never had the need to use my own prototypes. This pattern is meant to build objects by copy instead of constructors. Here is the definition given by the GoF:</p>
<p>&ldquo;Specify the kinds of objects to create using a prototypical instance, and create new objects by copying this prototype.&rdquo;</p>
<p>As often with the GoF, I don&rsquo;t understand their sentences (is this because English is not my native language?). If you&rsquo;re like me, here is another explanation: If you don&rsquo;t want or can&rsquo;t use the constructor of a class, the prototype pattern lets you create new instances of this class by duplicating an already existing instance.</p>
<p>Let&rsquo;s look at the formal definition using a UML diagram:</p>
<p><a href="http://coding-geek.com/wp-content/uploads/2015/06/prototype.png"><img alt="prototype pattern" src="http://coding-geek.com/pagespeed_static/1.JiBnMqyl6S.gif"></a></p>
<p>In this diagram</p>
<ul>
<li>the prototype is a an interface that defines a function clone()</li>
<li>A real prototype has&nbsp;to implement this interface and implement the clone() function to return an copy of itself.</li>
</ul>
<p>A developer will have to instantiate the ConcretePrototype once. Then, he will be able to create new instances of ConcretePrototype by:</p>
<ul>
<li>duplicating&nbsp;the first instance using the clone() function</li>
<li>or creating a ConcretePrototype using (again) the&nbsp;constructor.</li>
</ul>
<p>&nbsp;</p>
<h2><span id="When_to_use_prototypes">When to use prototypes?</span></h2>
<p>According to the Gof, the prototype should be used:</p>
<ul>
<li>when a system should be independent of how its products are created, composed, and represented</li>
</ul>
<ul>
<li>when the classes to instantiate are specified at run-time, for example, by dynamic loading</li>
</ul>
<ul>
<li>to avoid building a class hierarchy of factories that parallels the class hierarchy of products</li>
<li>when instances of a class can have one of only a few different combinations of state. It may be more convenient to install a corresponding number of prototypes and clone them rather than instantiating the class manually, each time with the appropriate state.</li>
</ul>
<p>&nbsp;</p>
<p>The dynamic loading of an unknown class is a very rare case, even more if the dynamically loaded instance needs to be duplicated.</p>
<p>This book was written in 1994. Now, you can&nbsp;&ldquo;avoid building a class hierarchy of factories&rdquo; by using dependency injection (again, I&rsquo;m going to present this wonderful pattern in a future article).</p>
<p>In my opinion the most common case is where creating a stateful instance is way more expensive than copying an existing instance and you need to create lots of this object. For example if the creation needs to:</p>
<ul>
<li>get data from a database connection,</li>
<li>get data from the system (with system calls) or the filesystem,</li>
<li>get data from another server (with sockets, web services or whatever),</li>
<li>compute a large amount of data (for example if it needs to sort data),</li>
<li>do anything that takes time.</li>
</ul>
<p>The object must be stateful because if it has no state, a Singleton (or a single instance) will do the trick.</p>
<p>There is also another use case. If you have an instance that is mutable and you want to give it to another part of the code, for security reasons you might want to give a duplicate instead of the real instance&nbsp;because this instance&nbsp;can be&nbsp;modified by the client code and have an impact on other parts of the code that use&nbsp;it.</p>
<p>&nbsp;</p>
<h2><span id="Javaimplementation">Java&nbsp;implementation</span></h2>
<p>Let&rsquo;s look at a simple example in Java:</p>
<ul>
<li>We&nbsp;have a CarComparator business class. This class contains a function that compares 2 cars.</li>
<li>To instanciate a&nbsp;CarComparator, the constructor needs to load a default configuration from a database to configure the car comparison algorithm (for example to put more weight on the fuel consumption than the speed or the price).</li>
<li>This class cannot be a singleton because the configuration can be modified by each user (therefore each user needs its own instance).</li>
<li>This is why we create only once a instance using the costly constructor.</li>
<li>Then when a client needs an instance of&nbsp;CarComparator he gets a duplicate of the first instance.</li>
</ul>
<pre>
//////////////////////////The Prototype interface
 public interface Prototype {
 Prototype duplicate();
 }
 
 //////////////////////////The class we want to duplicate
 public class CarComparator implements Prototype{
 private int priceWeigth;
 private int speedWeigth;
 private int fuelConsumptionWeigth;
 
 //a constructor that makes costly calls to a database
 //to get the default weigths
 public CarComparator(DatabaseConnection connect){
 //I let you imagine the costly calls to the database
 }
 
 //A private constructor only use to duplicate the object
 private CarComparator(int priceWeigth,int speedWeigth,int fuelConsumptionWeigth){
 this.priceWeigth=priceWeigth;
 this.speedWeigth=speedWeigth;
 this.fuelConsumptionWeigth=fuelConsumptionWeigth;
 }
 
 //The prototype method
 @Override
 public Prototype duplicate() {
 return new CarComparator(priceWeigth, speedWeigth, fuelConsumptionWeigth);
 }
 
 int compareCars(Car first, Car second){
 //some kickass and top secret algorithm using the weigths
 }
 
 ////////////////The setters that lets the possibility to modify
 //////////////// the algorithm behaviour
 public void setPriceWeigth(int priceWeigth) {
 this.priceWeigth = priceWeigth;
 }
 
 public void setSpeedWeigth(int speedWeigth) {
 this.speedWeigth = speedWeigth;
 }
 
 public void setFuelConsumptionWeigth(int fuelConsumptionWeigth) {
 this.fuelConsumptionWeigth = fuelConsumptionWeigth;
 }
 
 ////////////////////////// A factory that creates a CarComparator instance using
 ////////////////////////// constructors then it creates the others by duplication.
 ////////////////////////// When a client ask for a CarComparator
 ////////////////////////// he gets a duplicate
 
 public class CarComparatorFactory {
 CarComparator carComparator;
 public BusinessClass (DatabaseConnection connect) {
 //We create one instance of CarComparator
 carComparator = new CarComparator(connect);
 }
 
 //we duplicate the instance so that
 //the duplicated instances can be modified
 public CarComparator getCarComparator(){
 return carComparator.duplicate();
 }
 
 }
 
 
</pre>
<p>If you look at the next part, you&rsquo;ll see that I could have made a simpler code using the right Java interface but I wanted you to understand a prototype.</p>
<p>In this example, at start-up, the prototype will be created using the default configuration in the database and each client will get a copy of this instance using the getCarComparator() method of the factory.</p>
<p>&nbsp;</p>
<h2><span id="Real_example"><strong>Real example</strong></span></h2>
<p>The Java API provides a prototype interfaces called <a href="http://docs.oracle.com/javase/8/docs/api/java/lang/Cloneable.html">Cloneable</a>. This interface define a clone() function that a concrete prototype needs to implements. Many Java classes from the Java APIs implement this interface, for example the collections from the collection API. Using an ArrayList, I can clone it and get a new ArrayList that contains the same data as the original one:</p>
<pre>
 // Let's initialize a list
 // with 10 integers
 ArrayList&lt;Integer&gt; list = new ArrayList&lt;Integer&gt;();
 for (int i = 0; i &lt; 10; i++) {
 list.add(i);
 }
 System.out.println("content of the list "+list);
 
 //Let's now duplicate the list using the prototype method
 ArrayList&lt;Integer&gt; duplicatedSet = (ArrayList&lt;Integer&gt;) list.clone();
 System.out.println("content of the duplicated list "+duplicatedSet);
 
</pre>
<p>The result of this code is:</p>
<blockquote>
<p>content of the set [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]<br>
content of the duplicated set [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]</p>
</blockquote>
<p>&nbsp;</p>
<h1><span id="Builder_Pattern">Builder Pattern</span></h1>
<p>The Builder pattern is very useful to factorize code. According to the GoF, this pattern:</p>
<p><em>&ldquo;Separate the construction of a complex object from its representation so that the same construction process can create different representations.&rdquo;</em></p>
<p>Its aim has changed through time and&nbsp;it&rsquo;s most of the time used to avoid creating a lot of constructors that differs only by the number of arguments. It&rsquo;s a way to avoid the <strong>telescoping constructor anti-pattern</strong>.</p>
<p>&nbsp;</p>
<h2><span id="The_problem_it_solves">The problem it solves</span></h2>
<p>Let&rsquo;s look at the problem this pattern solves. Imagine a class Person with 5 attributes:</p>
<ul>
<li>age</li>
<li>weight</li>
<li>height</li>
<li>id</li>
<li>name</li>
</ul>
<p>We&nbsp;want to be able to construct a person knowing:</p>
<ul>
<li>only this age,</li>
<li>or only this age and weight,</li>
<li>or only this age, weight and height,</li>
<li>or only this age, weight, height and id</li>
<li>or only this age, weight, height, id and name</li>
</ul>
<p>&nbsp;</p>
<p>In java, we&nbsp;could write&nbsp;something like that</p>
<pre>
public class Person {
 private int age;
 private int weigth;
 private int height;
 private int id;
 private String name;
 
 //////////////Here comes the telescopic constructors
 public Person() {
 //some stuff
 }
 
 public Person(int age) {
 this();//we're using the previous constructor
 this.age = age;
 }
 
 public Person(int age, int weigth) {
 this(age);//we're using the previous constructor
 this.weigth = weigth;
 }
 
 public Person(int age, int weigth, int height) {
 this(age, weigth);//we're using the previous constructor
 this.height= height;
 } 
 
 public Person(int age, int weigth, int height,int id) {
 this(age, weigth, height);//we're using the previous constructor
 this.id = id;
 } 
 
 public Person(int age, int weigth, int height,int id,String name) {
 this(age, weigth, height, id);//we're using the previous constructor
 this.name = name;
 } 
 
 }
 
</pre>
<p>In order to deal with this simple need, we&rsquo;ve just created 5 constructors which a lot of code. I know java is a very verbose language (troll inside) but what if there was a cleaner way?</p>
<p>Moreover, using this telescopic approach, the code is hard to read. For example, if you read the following&nbsp;code, can you easily understand what the parameters are? Are they age, id or height?</p>
<pre>
Person person1 = new Person (45, 45, 160, 1);
 
 Person person2 = new Person (45, 170, 150);
 
 
</pre>
<p>Next problem: imagine you now want to be able to create a&nbsp;Person with every possible part of information you get, for example:</p>
<ul>
<li>an age</li>
<li>a weight</li>
<li>an age and a weight</li>
<li>an age and an id,</li>
<li>an age, a weight and a name,</li>
<li>&hellip;</li>
</ul>
<p>With the constructor approach, you&rsquo;ll end up with 120 constructors (5! = 120). And you&rsquo;ll have another problem, how can you deal with different constructors using the same types? For example how can you have&nbsp;both:</p>
<ul>
<li>a constructor for the age and the weigth (which are 2 int) and</li>
<li>a constructor for the age and the id (which are also 2 int)?</li>
</ul>
<p>You could use a static factory method&nbsp;but it would still require 120 static factory methods.</p>
<p><strong>This is where the builder comes into play!</strong></p>
<p>The idea of this pattern is to simulate <strong>named optional arguments</strong>. These types of arguments are natively available in some languages like python.</p>
<p>Since the UML version is very complicated (I think), we&rsquo;ll start with a simple java example and end with the UML formal definition.</p>
<p>&nbsp;</p>
<h2><span id="Asimple_java_example">A&nbsp;simple java example</span></h2>
<p>In this example, I have a person but this time the id field is mandatory and the other fields are optional.</p>
<p>I create a builder so that a developer can use the optional fields if he&nbsp;wants to.</p>
<pre>
//////////////////////the person class
 /////////////////////if you look at its constructor
 /////////////////////it requieres a builder
 public class Person {
 private final int id;// mandatory
 private int weigth;// optional
 private int height;// optional
 private int age;// optional
 private String name;// optional
 
 public Person(PersonBuilder builder) {
 age = builder.age;
 weigth = builder.weigth;
 height = builder.height;
 id = builder.id;
 name = builder.name;
 }
 }
 //////////////////////the builder that
 /////////////////////takes care of
 /////////////////////Person creation
 public class PersonBuilder {
 // Required parameters
 final int id;
 
 // Optional parameters - initialized to default values
 int height;
 int age;
 int weigth;
 String name = "";
 
 public PersonBuilder(int id) {
 this.id = id;
 }
 
 public PersonBuilder age(int val) {
 age = val;
 return this;
 }
 
 public PersonBuilder weigth(int val) {
 weigth = val;
 return this;
 }
 
 public PersonBuilder height(int val) {
 height = val;
 return this;
 }
 
 public PersonBuilder name(String val) {
 name = val;
 return this;
 }
 
 public Person build() {
 return new Person(this);
 }
 }
 
 //////////////////////Here is how to use the builder in order to build a Person
 //////////////////////You can see how readable is the code
 public class SomeClass {
 public void someMethod(int id){
 PersonBuilder pBuilder = new PersonBuilder(id);
 Person robert = pBuilder.name("Robert").age(18).weigth(80).build();
 //some stuff
 }
 
 public void someMethodBis(int id){
 PersonBuilder pBuilder = new PersonBuilder(id);
 Person jennifer = pBuilder.height(170).name("Jennifer").build();
 //some stuff
 }
 
 }
 
 
</pre>
<p>In this example, I suppose the classes Person and PersonBuilder&nbsp;are in the same package, which allows the builder to use the Person constructor and the&nbsp;classes outside the package will have to use the PersonBuilder to create a Person.</p>
<p>This PersonBuilder has 2 kinds of methods, one for <strong>building a part</strong> of a person and one for <strong>creating</strong> a person. All the properties of a person can only be modified by classes in the same package. &nbsp;I should have&nbsp;used getter and setter but I wanted to have a short example. You see that the part th uses the builder is easy to read, we know that we are creating</p>
<ul>
<li>a person named Robert whose age is 18 and weight 80,</li>
<li>another person named Jennifer whose length is&nbsp;170.</li>
</ul>
<p>Another advantage of this technic is that you can still create <strong>immutable objects</strong>. In my example, if I don&rsquo;t add public setters in the Person class, a&nbsp;Person instance&nbsp;is immutable since no class outside the package can modify its attributes.</p>
<p>&nbsp;</p>
<h2><span id="The_formal_definition">The formal definition</span></h2>
<p>Now let&rsquo;s look at the UML:</p>
<p><a href="http://coding-geek.com/wp-content/uploads/2015/06/builder.png"><img alt="builder pattern by the GoF" src="http://coding-geek.com/pagespeed_static/1.JiBnMqyl6S.gif"></a></p>
<p>&nbsp;</p>
<p>This diagram is really abstract, &nbsp;a GoF&rsquo;s&nbsp;builder has:</p>
<ul>
<li>a&nbsp;builder&nbsp;interface that specify functions&nbsp;for creating parts of a Product object. In my diagram, there is just one method, buildPart().</li>
<li>a ConcreteBuilder that constructs and assembles parts of the product by implementing the Builder interface.</li>
<li>a Director :&nbsp;it constructs a&nbsp;product&nbsp;using the Builder interface.</li>
</ul>
<p>&nbsp;</p>
<p>According to the GoF, this pattern is useful when:</p>
<blockquote>
<p>&bull; the algorithm for creating a complex object should be independent of the parts that make up the object and how they&rsquo;re assembled.<br>
&bull; the construction process must allow different representations for the object that&rsquo;s constructed.</p>
</blockquote>
<p>The example given by the GoF was a TextConverter builder that has 3 implementations to build: an ASCIIText&nbsp;or a&nbsp;TeXText or&nbsp;a TextWidget. The 3 builder implemenations (ASCIIConverter, TeXConverter and TextWidgetConverter) have the same functions except the createObject() function that differs (this is why this function is not in the interface of this pattern). Using this pattern, the code that converts a text (the Director) uses the builder interface so it can easily switch from ASCII to TeX or TextWidget. Moreover, you can add a new converter without modify the rest of the code. In a way, this pattern is very close to the State pattern.</p>
<p><strong>But this problem&nbsp;is a rare case</strong>.</p>
<p>&nbsp;</p>
<p>Another use of this pattern was popularized&nbsp;by Joshua Bloch, a Java developper who led the construction of many Java APIs. He wrote in his book <em>&ldquo;Effective Java&rdquo;</em>:</p>
<blockquote>
<p>&ldquo;Consider a builder when faced with many constructor parameters&rdquo;</p>
</blockquote>
<p><strong>Most of the time the pattern is used for this use case</strong>. You don&rsquo;t need the builder interface nor multiple builder implementations nor a director for this problem.&nbsp;In my java example and most of the time you will find just a&nbsp;concrete builder.</p>
<p>The UML then becomes easier:</p>
<p><a href="http://coding-geek.com/wp-content/uploads/2015/06/builder2.png"><img alt="builder pattern by Joshua Bloch" src="http://coding-geek.com/pagespeed_static/1.JiBnMqyl6S.gif"></a></p>
<p>In this diagram&nbsp;the ConcreteBuilder has multiple &nbsp;functions&nbsp; that create each part of the product (but I just put one, buildPart(), because I&rsquo;m lazy). These functions return the ConcreteBuilder so that you can chain the function calls,&nbsp;for example: builder.buildPart1().buildPart7().createObject(). The builder has a createObject() method to create the product when you don&rsquo;t need to add more parts.</p>
<p>&nbsp;</p>
<p>To sum up, the builder pattern is a good choice when&nbsp;you have a class with&nbsp;many optional parameters and you don&rsquo;t want to end up with to many constructors. Though this pattern was not designed&nbsp;for this problem, it&rsquo;s most of the time used for that (at least in Java).</p>
<p>&nbsp;</p>
<h2><span id="Real_example-2">Real example</span></h2>
<p>The most common&nbsp;example in the Java APIs is <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/StringBuilder.html">the StringBuilder</a>. Using it, you can create a temporary string, append new strings to it and when you&rsquo;re finished you can create a real String object (that is immutable).</p>
<pre>
 StringBuilder sBuilder = new StringBuilder();
 String example = sBuilder.append("this").append(" is").
 append(" an").append(" example").toString();
 System.out.println(example);
 
</pre>
<p>&nbsp;</p>
<h1><span id="Conclusion">Conclusion</span></h1>
<p>You should now have a better overview of the creational patterns. If you need to remember one thing it&rsquo;s to use single instances instead of singletons. Keep in mind the builder pattern (Joshua Bloch&rsquo;s version), it might be useful if you&rsquo;re dealing with optional parameters.</p>
</div>
</div>
<br>
<br>
<br>
<br>
Comp<br>
<br>
via coding-geek.com http://coding-geek.com<br>
<br>
April 19, 2019 at 12:20AM
